crates/ruff_python_trivia/src/tokenizer.rs

```diff
@@ -759,10 +757,12 @@ impl<'a> BackwardsTokenizer<'a> {
         Self {
             offset: range.start(),
             back_offset: range.end(),
-            // We could start tokenizing at a comment
+            // We only check for comments after a newline, so set to `true` to ensure we detect
+            // comments at the start of a range.
             after_newline: true,
-            comment_ranges: comment_range,
-            previous_comment_idx: None,
+            // Throw out any comments that follow the range.
+            comment_ranges: &comment_range
+                [..comment_range.partition_point(|comment| comment.start() <= range.end())],
```

Comment:
I must have missed this in the previous PR but it seems that the lexer now always initializes with `after_newline`. The lexer used to have a special `up_to` method indicating that it is guaranteed not to be after a newline to avoid testing for comments. 

I'm bringing this up here because the downside is that we now always perform a binary search to find the last comment, even if the caller can guarantee that the start position isn't after a newline. Bringing back this optimization is probably worth its own PR

Reply:
i didn't implement this initially because i thought about doing the binary search esp for large files with many comments is expensive to do for every is-expression-parenthesized check, but it doesn't seem to show in our benchmarks. Removing this is also kinda ugly because we have to bring back `after_newline`